### DISCUSSION TITLE
Custom distro got InaccessibleObjectException

### DISCUSSION THREAD

**Discussion by fangjian0423** [open-telemetry/opentelemetry-java-instrumentation#12318] [Q&A] [answered]
Hi team,

I have a custom distro, and will use URLClassLoader to load some 3rd libraries dynamically on the bootstrap phase.

Given `URLClassLoader#addURL` is a protected method, I should use reflection to get this method to invoke.

The code snippet like:

```java
URL url = file.toURI().toURL();
URLClassLoader.class.getDeclaredMethod("addURL", URL.class);
ADD_URL.setAccessible(true); // got error
addURL().invoke(classLoader, url);
```

I got below error when I use reflection:

```bash
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected void java.net.URLClassLoader.addURL(java.net.URL) accessible: module java.base does not "opens java.net" to unnamed module @32f1fafe
```

The workaround is adding a extra startup option `--add-opens=java.base/java.net=ALL-UNNAMED`.

I want to ask is there any ways to resolve it in otel-agent self programmatically?

Thank you!

**Comment by laurit** [comment, marked as answer]
See https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/faff2d995a1247f717cc0a4f25627efec1b178f6/muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentCachingPoolStrategy.java#L90-L120
